ReSharper.FSharp/src/FSharp.Psi.Features/src/Intentions/FunctionAnnotationAction.fs

```diff
@@ -1,59 +1,30 @@
 ﻿namespace JetBrains.ReSharper.Plugins.FSharp.Psi.Features.Intentions
 
 open FSharp.Compiler.SourceCodeServices
-open FSharp.Compiler.Layout
-open JetBrains.DocumentModel
 open JetBrains.ReSharper.Feature.Services.ContextActions
+open JetBrains.ReSharper.Plugins.FSharp.Psi.Impl.Tree
 open JetBrains.ReSharper.Plugins.FSharp.Psi.Parsing
 open JetBrains.ReSharper.Plugins.FSharp.Psi.Tree
 open JetBrains.ReSharper.Plugins.FSharp.Psi.Util
+open JetBrains.ReSharper.Plugins.FSharp.Services.Formatter
 open JetBrains.ReSharper.Psi.ExtensionsAPI
 open JetBrains.ReSharper.Psi.ExtensionsAPI.Tree
-open JetBrains.ReSharper.Psi.Tree
 open JetBrains.ReSharper.Resources.Shell
-open JetBrains.ReSharper.Plugins.FSharp
-
-type private ParameterToModify = {
-    ParameterNameNode : ILocalReferencePat
-    ParameterNodeInSignature : ISynPat
-}
+open JetBrains.ReSharper.Psi.Tree
+open JetBrains.Application.Settings
 
 [<ContextAction(Name = "AnnotateFunction", Group = "F#", Description = "Annotate function with parameter types and return type")>]
 type FunctionAnnotationAction(dataProvider: FSharpContextActionDataProvider) =
     inherit ContextActionBase()
 
     let [<Literal>] opName = "FunctionAnnotationAction"
-    
-    let getParameterTooltip (document: IDocument) (checkResults : FSharpCheckFileResults) pattern : string Option =
-        let offset = pattern.ParameterNameNode.GetTreeEndOffset().Offset
-        let coords = document.GetCoordsByOffset(offset)
-        let text = document.GetLineText(coords.Line)
-        let getTooltip =
-            checkResults.GetStructuredToolTipText(int coords.Line + 1, int coords.Column, text, [pattern.ParameterNameNode.SourceName], FSharpTokenTag.Identifier)
-        let (FSharpToolTipText layouts) = getTooltip.RunAsTask()
-        
-        let layout = layouts |> List.exactlyOne
-        match layout with
-        | FSharpStructuredToolTipElement.None
-        | FSharpStructuredToolTipElement.CompositionError _ -> failwith "Expected Group tooltip element"
-        | FSharpStructuredToolTipElement.Group(overloads) ->
-        let overload = overloads |> Seq.exactlyOne
-        
-        // Tooltips in this case are prepended with "val " which we don't want in the source code - in the case that
-        // a type can't be determined, it doesn't start with val, so don't try to annotate type
-        let tooltipString = showL overload.MainDescription
-        
-        if tooltipString.IndexOf("val ", 0) = 0 then
-            (showL overload.MainDescription |> Seq.skip 4 |> Seq.map string) |> String.concat "" |> Some
-        else
-            None
     override x.Text = "Annotate function with parameter types and return type"
     
     override x.IsAvailable _ =
         let binding = dataProvider.GetSelectedElement<IBinding>()
         if isNull binding then false else
 
-        match binding.HeadPattern.As<INamedPat>() with
+        match binding.HeadPattern.As<IParametersOwnerPat>() with
```

Comment:
I think I'd be OK to do something like
```fsharp
match binding.HeadPattern with
| :? IAsPat -> false
| :? INamedPat as namedPat -> isNotNull namedPat.identifer
| _ -> false
```

It'd allow annotating values, as before this last change, like in
```fsharp
let v = 123
```